libflux/src/semantic/nodes.rs

```diff
@@ -542,8 +572,61 @@ pub struct BinaryExpr {
 }
 
 impl BinaryExpr {
-    fn infer(&self, env: Environment, f: &mut Fresher) -> Result {
-        unimplemented!();
+    fn infer(&mut self, env: Environment, f: &mut Fresher) -> Result {
+        // Compute the left and right constraints.
+        // Do this first so that we can return an error if one occurs.
+        let (env, lcons) = self.left.infer(env, f)?;
+        let (env, rcons) = self.right.infer(env, f)?;
+
+        let mut cons = Vec::new();
+        if let Some(kind) = match self.operator {
+            ast::Operator::AdditionOperator => Some(Addable),
+            ast::Operator::SubtractionOperator => Some(Subtractable),
+            _ => None,
+        } {
+            cons.push(Constraint::Kind(self.typ.clone(), kind));
+        }
+
+        cons.append(&mut Vec::from(match self.operator {
+            // The following operators require both sides to be equal.
+            ast::Operator::AdditionOperator
+            | ast::Operator::SubtractionOperator
+            | ast::Operator::MultiplicationOperator
+            | ast::Operator::DivisionOperator
```

Comment:
Including multiplication?

Reply:
Yes, I could be assuming wrong, but any type that can be an operand of `*` can also be an operand of `/`

Reply:
Just noting for the future this might not be the case. Or at the minimum, we will likely want to revisit this because this same principle would probably apply to addition and subtraction too.